python/ray/autoscaler/_private/gcp/tpu_command_runner.py

```diff
@@ -0,0 +1,164 @@
+"""Command runners specific to TPU VM pods.
+
+TPU VM pods may contain multiple hosts, each including attached TPU chips and
+associated internal/external IP addresses.
+
+To support TPU VM pods, we represent entire TPU pods as "Ray Nodes", meaning
+that TPU pods will need to run the operations specified in `CommandRunnerInterface`
+N times, where N denotes the number of hosts that comprise a TPU pod.
+
+To maintain feature completeness, we simply wrap the existing `SSHCommandRunner` and
+`DockerCommandRunner` and run them as batched calls.
+
+"""
+from concurrent.futures import ThreadPoolExecutor
+from types import ModuleType
+from typing import Any, Dict, Optional
+
+from ray.autoscaler._private.command_runner import DockerCommandRunner, SSHCommandRunner
+from ray.autoscaler._private.gcp.node import GCPTPUNode
+from ray.autoscaler.command_runner import CommandRunnerInterface
+from ray.autoscaler.node_provider import NodeProvider
+
+# We see issues where too many concurrent connections may lead to failed SSH connections
+# when there are too many TPU hosts.
+# For now, we cap the maximum concurrent connections until we find the right fix.
+_MAX_NUM_CONCURRENT_ACTIVE_CONNECTIONS = 16
```

Comment:
Does it make sense to make this configurable via an environment variable?

Reply:
That makes sense, any suggestions on a name?

Reply:
Mm, maybe `RAY_TPU_MAX_CONCURRENT_ACTIVE_CONNECTIONS`?

Reply:
Added! Btw, is there a way to set this value within the cluster yaml file? I couldn't find one, but that would be great for usability.

Reply:
I don't think so unfortunately, there's currently no place to set env vars in the cluster YAML that would be read by the cluster launcher code itself.